### PR TITLE
N distinct eval tidy

### LINKED ISSUE #5069: n_distinct() errors when passed two or more vectors of same type
In the current dev version of dplyr (0.8.99.9002), the function `n_distinct()` throws an error when the `...` arguments include two or more vectors having the same type. For example:

```r
n_distinct(iris$Sepal.Length, iris$Sepal.Width)
# Error: Column name `<dbl>` must not be duplicated.
```

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #5069

``` r
dplyr::n_distinct(iris$Sepal.Length, iris$Sepal.Width)
#> [1] 117
```

<sup>Created on 2020-03-30 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### CODE REVIEW

File: R/distinct.R [hunk 1]
```diff
@@ -136,8 +137,8 @@ distinct.data.frame <- function(.data, ..., .keep_all = FALSE) {
 #' n_distinct(x)
 #' @export
 n_distinct <- function(..., na.rm = FALSE) {
-  # TODO: remove !!!list2() when tibble abandons `.data` masking
-  data <- tibble(!!!list2(...))
+  columns <- map(enquos(..., .named = TRUE), eval_tidy)
+  data <- tibble(!!!columns)
```
Comment (hadley) [inline R/distinct.R]: Maybe `vctrs::new_data_frame()`?

Comment (romainfrancois) [reply R/distinct.R]: what `tibble(!!!)` gives is tidy recycling, which is tested : 

```r
test_that("n_distinct recycles length 1 vectors (#3685)", {
  expect_equal(n_distinct(1, 1:4), 4)
  expect_equal(n_distinct(1:4, 1), 4)

  d <- tibble(x = 1:4)
  res <- d %>%
    summarise(y = sum(x), n1 = n_distinct(y, x), n2 = n_distinct(x, y), n3 = n_distinct(y), n4 = n_distinct(identity(y)), n5 = n_distinct(x))
  expect_equal(res$n1, 4)
  expect_equal(res$n2, 4)
  expect_equal(res$n3, 1)
  expect_equal(res$n4, 1)
  expect_equal(res$n5, 4)

  res <- tibble(g = c(1,1,1,1,2,2), x = c(1,2,3,1,1,2)) %>%
    group_by(g) %>%
    summarise(y = sum(x), n1 = n_distinct(y, x), n2 = n_distinct(x, y), n3 = n_distinct(y), n4 = n_distinct(identity(y)), n5 = n_distinct(x))
  expect_equal(res$n1, c(3,2))
  expect_equal(res$n2, c(3,2))
  expect_equal(res$n3, c(1,1))
  expect_equal(res$n4, c(1,1))
  expect_equal(res$n5, c(3,2))
})
```

Comment (krlmlr) [reply R/distinct.R]:
```suggestion
  data <- as_tibble(columns, .name_repair = "minimal")
```

Comment (romainfrancois) [reply R/distinct.R]: Thanks

**Pull request merged**